tests/lib/ocpp/v201/functional_blocks/test_reservation.cpp

```diff
@@ -0,0 +1,593 @@
+    }
+
+    ///
+    /// \brief Create example CancelReservation request to use in tests.
+    /// \param reservation_id   The reservation id.
+    /// \return The request message.
+    ///
+    ocpp::EnhancedMessage<MessageType> create_example_cancel_reservation_request(const int32_t reservation_id) {
+        CancelReservationRequest request;
+        request.reservationId = reservation_id;
+        ocpp::Call<CancelReservationRequest> call(request);
+        ocpp::EnhancedMessage<MessageType> enhanced_message;
+        enhanced_message.messageType = MessageType::CancelReservation;
+        enhanced_message.message = call;
+        return enhanced_message;
+    }
+
+protected: // Members
+    // DatabaseConnection as member so the database keeps open and is not destroyed (because this is an in memory
+    // database).
+    std::unique_ptr<ocpp::common::DatabaseConnection> database_connection;
+    MockMessageDispatcher mock_dispatcher;
+    EvseManagerFake evse_manager{NR_OF_EVSES};
+    // Device model is a unique ptr here because of the database: it is stored in memory so as soon as the handle to
+    // the database closes, the database is removed. So the handle should be opened before creating the devide model.
+    // So the device model is initialized on nullptr, then the handle is opened, the devide model is created and the
+    // handle stays open until the whole test is destructed.
+    std::unique_ptr<DeviceModel> device_model;
+    MockFunction<ReserveNowStatusEnum(const ReserveNowRequest& request)> reserve_now_callback_mock;
+    MockFunction<bool(const int32_t reservationId)> cancel_reservation_callback_mock;
+    MockFunction<ocpp::ReservationCheckStatus(const int32_t evse_id, const ocpp::CiString<36> idToken,
+                                              const std::optional<ocpp::CiString<36>> groupIdToken)>
+        is_reservation_for_token_callback_mock;
+    // Make reservation a unique ptr so we can create it after creating the device model.
+    std::unique_ptr<Reservation> reservation;
+
+    IdToken default_test_token;
+};
+
+TEST_F(ReservationTest, handle_reserve_now_reservation_not_available) {
+    // In the device model, reservation is set to not available. This should reject the request.
+    set_reservation_available(this->device_model.get(), false);
+
+    EXPECT_CALL(mock_dispatcher, dispatch_call_result(_)).WillOnce(Invoke([](const json& call_result) {
+        auto response = call_result[ocpp::CALLRESULT_PAYLOAD].get<ReserveNowResponse>();
+        EXPECT_EQ(response.status, ReserveNowStatusEnum::Rejected);
+        ASSERT_TRUE(response.statusInfo.has_value());
+        ASSERT_TRUE(response.statusInfo.value().additionalInfo.has_value());
+        EXPECT_EQ(response.statusInfo.value().additionalInfo.value(), "Reservation is not available");
+    }));
+
+    EvseMock& m1 = evse_manager.get_mock(1);
+    EvseMock& m2 = evse_manager.get_mock(2);
+
+    EXPECT_CALL(m1, get_connector_status(_)).Times(0);
```

Comment:
I think these are also not really relevant. Its already improved by quite a bit at least so would be ok to leave these for now.

Reply:
Yes i wanted to test if it is really doing nothing there. Only directly returns.